x/stake/keeper/slash_test.go

```diff
@@ -0,0 +1,327 @@
+	keeper.SetRedelegation(ctx, rd)
+	slashAmount, _ = keeper.slashRedelegation(ctx, rd, 0, fraction)
+	require.Equal(t, int64(5), slashAmount.Evaluate())
+	rd, found := keeper.GetRedelegation(ctx, addrDels[0], addrVals[0], addrVals[1])
+	require.True(t, found)
+	// initialbalance unchanged
+	require.Equal(t, sdk.NewCoin(params.BondDenom, 10), rd.InitialBalance)
+	// balance decreased
+	require.Equal(t, sdk.NewCoin(params.BondDenom, 5), rd.Balance)
+
+	// shares decreased
+	del, found = keeper.GetDelegation(ctx, addrDels[0], addrVals[1])
+	require.True(t, found)
+	require.Equal(t, int64(5), del.Shares.Evaluate())
+}
+
+// tests Slash at a future height (must panic)
+func TestSlashAtFutureHeight(t *testing.T) {
+	ctx, keeper, _, _, pk := setupHelper(t, 10)
+	fraction := sdk.NewRat(1).Quo(sdk.NewRat(2))
+	require.Panics(t, func() { keeper.Slash(ctx, pk, 1, 10, fraction) })
+}
+
+// tests Slash at the current height
+func TestSlashAtCurrentHeight(t *testing.T) {
+	ctx, keeper, _, _, pk := setupHelper(t, 10)
+	fraction := sdk.NewRat(1).Quo(sdk.NewRat(2))
+
+	oldPool := keeper.GetPool(ctx)
+	validator, found := keeper.GetValidatorByPubKey(ctx, pk)
+	require.True(t, found)
+	keeper.Slash(ctx, pk, ctx.BlockHeight(), 10, fraction)
+
+	// read updated state
+	validator, found = keeper.GetValidatorByPubKey(ctx, pk)
+	require.True(t, found)
+	newPool := keeper.GetPool(ctx)
+
+	// power decreased
+	require.Equal(t, sdk.NewRat(5), validator.GetPower())
+	// pool bonded shares decreased
+	require.Equal(t, sdk.NewRat(5).Evaluate(), oldPool.BondedShares.Sub(newPool.BondedShares).Evaluate())
+}
+
+// tests Slash at a previous height with an unbonding delegation
+func TestSlashWithUnbondingDelegation(t *testing.T) {
+	ctx, keeper, params, _, pk := setupHelper(t, 10)
+	fraction := sdk.NewRat(1).Quo(sdk.NewRat(2))
+
+	// set an unbonding delegation
+	ubd := types.UnbondingDelegation{
+		DelegatorAddr:  addrDels[0],
+		ValidatorAddr:  addrVals[0],
+		CreationHeight: 11,
+		MinTime:        0,
+		InitialBalance: sdk.NewCoin(params.BondDenom, 4),
+		Balance:        sdk.NewCoin(params.BondDenom, 4),
+	}
+	keeper.SetUnbondingDelegation(ctx, ubd)
+
+	// slash validator
+	ctx = ctx.WithBlockHeight(12)
+	oldPool := keeper.GetPool(ctx)
+	validator, found := keeper.GetValidatorByPubKey(ctx, pk)
+	require.True(t, found)
+	keeper.Slash(ctx, pk, 10, 10, fraction)
+
+	// read updating unbonding delegation
+	ubd, found = keeper.GetUnbondingDelegation(ctx, addrDels[0], addrVals[0])
+	require.True(t, found)
+	// balance decreased
+	require.Equal(t, sdk.NewInt(2), ubd.Balance.Amount)
+	// read updated pool
+	newPool := keeper.GetPool(ctx)
+	// bonded tokens burned
+	require.Equal(t, int64(3), oldPool.BondedTokens-newPool.BondedTokens)
+	// read updated validator
+	validator, found = keeper.GetValidatorByPubKey(ctx, pk)
+	// power decreased, but not by quite half, stake was bonded since
+	require.Equal(t, sdk.NewRat(7), validator.GetPower())
```

Comment:
confusing test to understand - decrease by not quite half ? should outline which coins are being removed from where - I still don't really understand where 7 comes from

Reply:
Added a comment - this is right, it is a result of how much stake bonded at the time of infraction is still bonded at the time of discovery - but I'm also going to add more tests with less implicit events.